vortex-serde/src/layouts/pruning.rs

```diff
@@ -0,0 +1,538 @@
+        self.stat
+    }
+}
+
+#[allow(dead_code)]
+pub struct PruningPredicate {
+    expr: Arc<dyn VortexExpr>,
+    stat_references: HashMap<Field, ColumnStat>,
+}
+
+impl PruningPredicate {
+    pub fn new(original_expr: &Arc<dyn VortexExpr>) -> Self {
+        let (expr, stat_references) = convert_to_pruning_expression(original_expr);
+        Self {
+            expr,
+            stat_references,
+        }
+    }
+
+    pub fn references(&self) -> &HashMap<Field, ColumnStat> {
+        &self.stat_references
+    }
+
+    pub fn necessary_statistics(&self) -> HashMap<Field, Vec<Stat>> {
+        let mut to_fetch: HashMap<Field, Vec<Stat>> = HashMap::new();
+        for column_stat in self.stat_references.values() {
+            match to_fetch.entry(column_stat.column().clone()) {
+                Entry::Occupied(o) => o.into_mut().push(column_stat.stat()),
+                Entry::Vacant(v) => {
+                    v.insert(vec![column_stat.stat()]);
+                }
+            }
+        }
+        to_fetch
+    }
+}
+
+fn convert_to_pruning_expression(
+    expr: &Arc<dyn VortexExpr>,
+) -> (Arc<dyn VortexExpr>, HashMap<Field, ColumnStat>) {
+    // Anything that can't be translated has to be represented as
+    // boolean true expression, i.e. the value might be in that chunk
+    let fallback = Arc::new(Literal::new(Scalar::bool(true, Nullability::NonNullable)));
+    // TODO(robert): Add support for boolean column expressions,
+    //  i.e. if column is of bool dtype it's valid to filter on it directly as a predicate
+    if expr.as_any().downcast_ref::<Column>().is_some() {
+        return (fallback, HashMap::new());
+    }
+
+    if let Some(bexp) = expr.as_any().downcast_ref::<BinaryExpr>() {
+        if bexp.op() == Operator::Or || bexp.op() == Operator::And {
+            let (rewritten_left, mut refs_lhs) = convert_to_pruning_expression(bexp.lhs());
+            let (rewritten_right, refs_rhs) = convert_to_pruning_expression(bexp.rhs());
+            refs_lhs.extend(refs_rhs);
+            return (
+                Arc::new(BinaryExpr::new(rewritten_left, bexp.op(), rewritten_right)),
+                refs_lhs,
+            );
+        }
+
+        if let Some(col) = bexp.lhs().as_any().downcast_ref::<Column>() {
+            return PruningPredicateBuilder::try_new(col.field().clone(), bexp.op(), bexp.rhs())
+                .and_then(column_comparison_to_stat_expr)
+                .unwrap_or_else(|| (fallback, HashMap::new()));
+        };
+
+        if let Some(col) = bexp.rhs().as_any().downcast_ref::<Column>() {
+            return PruningPredicateBuilder::try_new(
+                col.field().clone(),
+                bexp.op().swap(),
+                bexp.lhs(),
+            )
+            .and_then(column_comparison_to_stat_expr)
+            .unwrap_or_else(|| (fallback, HashMap::new()));
+        };
+    }
+
+    (fallback, HashMap::new())
+}
+
+struct PruningPredicateBuilder<'a> {
```

Comment:
`PruningPredicateRewriter`? Its not really used as a builder IMO

Reply:
yes, I struggled with a good name